source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+     play(tx0, p, driveframe1);
+     delay[20dt] driveframe1;
+     // Clocks now unaligned by 120dt, so we use a `barrier` to re-align
+     barrier(driveframe1, driveframe2);
+     // `driveframe2` will now play a pulse 120dt after `driveframe1` finishes playing
+     play(tx0, p, driveframe2);
+   }
+
+Examples
+--------
+
+Cross-resonance gate
+~~~~~~~~~~~~~~~~~~~~
+
+
+.. code-block:: javascript
+
+  cal {
+     frame frame0 = newframe(5.0e9, 0);
+  }
+
+  defcal cross_resonance $0 $1 {
+      // Access globally defined channels
+      channel d0 = txch($0, "drive");
+      channel d1 = txch($1, "drive");
+
+      waveform wf1 = gaussian_square(1., 1024dt, 128dt, 32dt);
+      waveform wf2 = gaussian_square(0.1, 1024dt, 128dt, 32dt);
+
+      // phase update some virtual Z gate
+      frame0.phase += pi/2;
+
+      /*** Do pre-rotation ***/
+      {...}
+
+      // generate new frame for second drive that is locally scoped
+      frame temp_frame = copyframe(frame0);
+      temp_frame.phase = frame0.phase + pi/2;
+
+      play(d0, wf1, frame0);
+      play(d1, wf2, temp_frame);
+
+      /*** Do post-rotation ***/
+      {...}
+  }
+
+Geometric gate
+~~~~~~~~~~~~~~
+
+.. code-block:: javascript
+
+  float[32] fq_01 = 5e9; // hardcode or pull from some function
+  float[32] anharm = 300e6; // hardcode or pull from some function
+  cal {
+      frame frame_01 = newframe(fq_01, 0);
+      frame frame_12 = newframe(fq_01 + anharm, 0);
+  }
+
+  defcal geo_gate(angle[32] theta) $0 {
+      // theta: rotation angle (about z-axis) on Bloch sphere
+
+      // Access globally defined channels
+      tx_channel dq = txch($q, “drive”);
+
+      // Assume we have calibrated 0->1 pi pulses and 1->2 pi pulse
+      // envelopes (no sideband)
+      waveform X_01 = {...};
+      waveform X_12 = {...};
+      float[32] a = sin(theta/2);
+      float[32] b = sqrt(1-a**2);
+
+      // Double-tap
+      play(dq, scale(a, X_01), frame_01);
+      play(dq, scale(b, X_12), frame_12);
+      play(dq, scale(a, X_01), frame_01);
+      play(dq, scale(b, X_12), frame_12);
+  }
+
+Neutral atoms
+~~~~~~~~~~~~~
+
+In this simple example, the signal chain is composed of two electro-optic modulators (EOM) and
+an acousto-optic deflector (AOD). The EOMs put sidebands on the laser light while the AOD diffracts
+the light in an amount proportional to the frequency of the RF drive. This example was chosen
+because it is similar in spirit to the work by Levine et al.:cite:`levine2019` except that phase
+control is exerted using virtual Z gates on the AODs -- requiring frame tracking of the qubit
+frequency yet application of a tone that maps to the qubit position (i.e. requires the use of a
+sideband).
+
+The program aims to perform a Hahn echo sequence on q1, and a Ramsey sequence on q2 and q3.
+
+.. code-block:: javascript
+
+  defcal neutral_atoms {
+    // Access globally defined channels
+    channel eom_a_channel = txch(0, "eom_a");
```

Comment:
Make these `txchannel`'s.

Reply:
Also should they be physical qubits? Ie `0`->`$0`, et...

Reply:
I think the agreement offline was to use `channel` . Also, yes they should be physical qubits.

Reply:
0f5972b